File: python/cadctl/analysis_model.py
"""Analysis-model derivation records (0.8 review P0-6).

The 0.8 analysisModel guard proved the SOURCE was canonical, but not that
the derived geometry actually came from it: any unrelated STEP could claim
provenance and re-bind its evidence to the real design. This module makes
derivations harness-owned records instead of agent claims:

    fused | bonded   the harness performs the boolean union itself and
                     writes the output STEP — the derivation is mechanical
                     and fully verified;
    simplified |
    defeatured |
    sectioned        the Agent authors the model; the harness hashes both
                     ends at record time (authored: true). This proves the
                     chain was created through a harness tool against the
                     canonical source at a specific time — weaker than a
                     mechanical derivation, and labeled as such.

Simulations accept only { derivationRef } pointing at a stored record; the
guard re-verifies record.sourceHash is canonical and record.outputHash
matches the geometry actually being solved.
"""

from __future__ import annotations

import json
import time
from pathlib import Path
from typing import Any

from .common import sha256_file

OPERATIONS_MECHANICAL = {"fused", "bonded"}
OPERATIONS_AUTHORED = {"simplified", "defeatured", "sectioned"}
ALL_OPERATIONS = OPERATIONS_MECHANICAL | OPERATIONS_AUTHORED

RECORD_SCHEMA_VERSION = 1


def validate_derive_spec(spec: dict[str, Any]) -> tuple[bool, list[str]]:
    errors: list[str] = []
    if not isinstance(spec, dict):
        return False, ["spec must be an object"]
    unknown = sorted(set(spec) - {"source", "operations", "output"})
    if unknown:
        errors.append(f"spec has unknown keys {unknown}; allowed keys are ['operations', 'output', 'source']")

    source = spec.get("source")
    if not isinstance(source, str) or not source.strip():
        errors.append("source is required (the authoritative design)")
    elif Path(source).suffix.lower() not in (".step", ".stp"):
        errors.append("source must be .step or .stp")
    elif not Path(source).is_file():
        errors.append(f"source does not exist: {source}")

    operations = spec.get("operations")
    if not isinstance(operations, list) or not operations:
        errors.append("operations must be a non-empty list")
    else:
        for op in operations:
            if op not in ALL_OPERATIONS:
                errors.append(f"operations entries must be one of {sorted(ALL_OPERATIONS)}; got {op!r}")
        mechanical = [op for op in operations if op in OPERATIONS_MECHANICAL]
        authored = [op for op in operations if op in OPERATIONS_AUTHORED]
        if mechanical and authored:
            errors.append("mechanical operations (fused/bonded) cannot be combined with authored ones (simplified/defeatured/sectioned)")

    output = spec.get("output")
    if output is not None and not isinstance(output, str):
        errors.append("output must be a path string")

    return not errors, errors


def run_derivation(spec_path: str | Path, output_dir: str | Path) -> dict[str, Any]:
    started = time.monotonic()
    spec_path = Path(spec_path)
    spec = json.loads(spec_path.read_text(encoding="utf-8"))
    ok, errors = validate_derive_spec(spec)
    if not ok:
        raise ValueError("; ".join(errors))

    output_dir = Path(output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)
    source = Path(spec["source"]).resolve()
    operations = list(spec["operations"])
    mechanical = any(op in OPERATIONS_MECHANICAL for op in operations)
    source_hash = sha256_file(source)

    if mechanical:
        # The harness performs the union itself: the derivation is exactly
        # reproducible and cannot be an unrelated model in disguise.
        output = Path(spec.get("output") or output_dir / "analysis-model.step").resolve()
        with _union_part(source) as union:
            import build123d as bd

            bd.export_step(union, output)
        executed = True
    else:
        # Authored derivation: the Agent's model is hashed at record time.
        if not spec.get("output"):
            raise ValueError("authored operations require output (the model you authored)")
        output = Path(spec["output"]).resolve()
        if not output.is_file():
            raise ValueError(f"authored output does not exist: {output}")
        if output.resolve() == source.resolve():
            raise ValueError("authored output must differ from the source (use the canonical artifact directly instead)")
        executed = False

    output_hash = sha256_file(output)
    record: dict[str, Any] = {
        "schemaVersion": RECORD_SCHEMA_VERSION,
        "source": str(source),
        "sourceHash": source_hash,
        "operations": operations,
        "output": str(output),
        "outputHash": output_hash,
        "executed": executed,
        "notes": (
            "harness-executed boolean union: derivation is mechanically verified"
            if executed
            else "agent-authored model: the harness records the hash chain at creation time; the simplification itself is the Agent's modeling judgment"
        ),
        "durationMs": int((time.monotonic() - started) * 1000),
    }
    record_path = output_dir / "derivation.json"
    record_path.write_text(json.dumps(record, indent=2), encoding="utf-8")
    record["recordPath"] = str(record_path)
    return record


class _union_part:
    """Context manager producing the boolean union of a STEP's solids."""

    def __init__(self, source: Path):
        self.source = source

    def __enter__(self):
        import build123d as bd
        from OCP.BRepAlgoAPI import BRepAlgoAPI_Fuse
        from OCP.TopTools import TopTools_ListOfShape

        shape = bd.import_step(self.source)
        solids = list(shape.solids())
        if not solids:
            raise ValueError("source contains no solids to fuse")
        result = solids[0].wrapped
        for solid in solids[1:]:
            args = TopTools_ListOfShape()
            args.Append(result)
            tools = TopTools_ListOfShape()
            tools.Append(solid.wrapped)
            fuse = BRepAlgoAPI_Fuse()
            fuse.SetArguments(args)
            fuse.SetTools(tools)
            fuse.SetRunParallel(False)
            fuse.Build()
            if not fuse.IsDone():
                raise RuntimeError("boolean fuse failed: derivation unresolved")
            result = fuse.Shape()
        return bd.Part(result)

    def __exit__(self, *exc):
        return False

File: src/extensions/simulation/analysis-model.ts
/**
 * Authoritative vs analysis model (0.8 M4 + review P0-6).
 *
 * The canonical design is never fused, bonded, or simplified for solver
 * convenience. When a simulation's geometry input is a DERIVED model, the
 * derivation must be a harness-owned record, not an agent claim:
 *
 *   cad_derive_analysis_model({ source, operations, output? })
 *     - fused/bonded: the harness performs the boolean union itself and
 *       writes the output STEP — mechanically verified.
 *     - simplified/defeatured/sectioned: the Agent authors the model; the
 *       harness hashes both ends at record time (authored, labeled).
 *
 * Simulations declare { analysisModel: { derivationRef } } pointing at the
 * stored record. The guard verifies:
 *   1. the record exists and parses;
 *   2. record.sourceHash IS a canonical artifact of the workspace
 *      (compared by hash only — no geometry heuristics);
 *   3. record.outputHash MATCHES the geometry actually being solved —
 *      an unrelated STEP can no longer claim provenance, closing the
 *      "source is real but the derivation is not" hole.
 *
 * The evidence SUBJECT then becomes the authoritative design; the derived
 * model stays a hash-bound input.
 */

import { existsSync } from "node:fs";
import { readFile } from "node:fs/promises";
import { resolve } from "node:path";
import { Type } from "typebox";

import { CadProjectStore, sha256File } from "../../shared/store.ts";

export const DeriveAnalysisModelSchema = Type.Object(
  {
    source: Type.String({ description: "Authoritative design STEP this model derives from" }),
    operations: Type.Array(
      Type.Enum({
        fused: "fused",
        bonded: "bonded",
        simplified: "simplified",
        defeatured: "defeatured",
        sectioned: "sectioned",
      }),
      { minItems: 1, description: "fused/bonded are executed by the harness; simplified/defeatured/sectioned record your authored model" },
    ),
    output: Type.Optional(
      Type.String({ description: "Output STEP. Required for authored operations; optional for fused/bonded (harness writes it)" }),
    ),
  },
  { additionalProperties: false },
);

export const AnalysisModelSchema = Type.Object(
  {
    derivationRef: Type.String({
      description: "Path of the harness-owned derivation record (from cad_derive_analysis_model)",
    }),
  },
  { additionalProperties: false },
);

export interface AnalysisModelDeclaration {
  derivationRef: string;
}

interface DerivationRecord {
  sourceHash?: string;
  outputHash?: string;
  executed?: boolean;
  operations?: string[];
}

export interface AnalysisModelCheck {
  /** Fail-closed reason; absent when the subject is legitimate. */
  error?: string;
  /**
   * Evidence subject override: the authoritative design's hash when the
   * subject is a derived model with a valid derivation record.
   */
  subjectOverrideHash?: string;
}

async function canonicalHashes(cwd: string): Promise<Set<string>> {
  const canonical = new Set<string>();
  const store = new CadProjectStore(cwd);
  try {
    const state = await store.load();
    if (state?.currentArtifactHash) canonical.add(state.currentArtifactHash);
    if (state?.baselineArtifactHash) canonical.add(state.baselineArtifactHash);
  } catch {
    // no active run — the head still defines the design
  }
  try {
    const project = await store.loadProject();
    if (project?.head.artifactHash) canonical.add(project.head.artifactHash);
  } catch {
    // no project yet (adhoc use) — nothing to protect
  }
  return canonical;
}

/**
 * Verify the simulation subject against the canonical design hashes of the
 * workspace. Rules:
 *
 *   1. subject hash is canonical -> fine (derived inputs like fluidDomain
 *      are already hash-bound via inputArtifacts);
 *   2. subject is derived with a valid derivation record -> fine, and the
 *      evidence subject becomes the record's source (which must itself be
 *      canonical);
 *   3. subject is derived with no declaration -> fail closed;
 *   4. no canonical design in the workspace -> adhoc use, nothing to
 *      protect.
 */
export async function verifyAnalysisModel(
  cwd: string,
  opts: {
    subject: string;
    analysisModel?: AnalysisModelDeclaration;
  },
): Promise<AnalysisModelCheck> {
  const canonical = await canonicalHashes(cwd);
  if (canonical.size === 0) return {};
  const subjectAbs = resolve(cwd, opts.subject);
  if (!existsSync(subjectAbs)) return { error: `simulation subject does not exist: ${opts.subject}` };
  const subjectHash = await sha256File(subjectAbs);
  if (canonical.has(subjectHash)) return {};

  if (!opts.analysisModel) {
    return {
      error: [
        `simulation subject ${opts.subject} is not the canonical design of this workspace`,
        "and no analysisModel is declared: the canonical design is never fused/bonded/simplified for solver convenience",
        "create a derivation with cad_derive_analysis_model, then declare analysisModel {derivationRef}",
      ].join("; "),
    };
  }
  const recordAbs = resolve(cwd, opts.analysisModel.derivationRef);
  if (!existsSync(recordAbs)) {
    return { error: `analysisModel.derivationRef does not exist: ${opts.analysisModel.derivationRef}` };
  }
  let record: DerivationRecord;
  try {
    record = JSON.parse(await readFile(recordAbs, "utf-8")) as DerivationRecord;
  } catch {
    return { error: `analysisModel.derivationRef is not a readable derivation record: ${opts.analysisModel.derivationRef}` };
  }
  if (!record.sourceHash || !record.outputHash) {
    return { error: `derivation record is malformed (needs sourceHash and outputHash): ${opts.analysisModel.derivationRef}` };
  }
  // THE closing check: the geometry being solved must be exactly what the
  // derivation produced — an unrelated STEP cannot borrow provenance.
  if (record.outputHash !== subjectHash) {
    return {
      error: [
        `simulation subject ${opts.subject} is not the model the derivation record produced`,
        "(record output hash does not match the subject): create a fresh derivation for this geometry",
      ].join("; "),
    };
  }
  if (!canonical.has(record.sourceHash)) {
    return {
      error: `derivation record's source is not a canonical design artifact of this workspace`,
    };
  }
  // The evidence is ABOUT the authoritative design; the derived model is
  // only a hash-bound input.
  return { subjectOverrideHash: record.sourceHash };
}
